plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/WarPublisher.java

```diff
@@ -51,40 +52,40 @@
    * directory already contains resources those will be deleted if they are not part of the exploded
    * WAR.
    */
-  public static void publishExploded(IProject project, IPath destination, IProgressMonitor monitor)
-      throws CoreException {
-    publish(project, destination, true /* exploded */, monitor);
+  public static void publishExploded(IProject project, IPath destination,
+      IPath safeWorkDirectory, IProgressMonitor monitor) throws CoreException {
+    publish(project, destination, safeWorkDirectory, true /* exploded */, monitor);
   }
 
-  public static void publishWar(IProject project, IPath destination, IProgressMonitor monitor)
-      throws CoreException {
-    publish(project, destination, false /* exploded */, monitor);
+  public static void publishWar(IProject project, IPath destination, IPath safeWorkDirectory,
+      IProgressMonitor monitor) throws CoreException {
+    publish(project, destination, safeWorkDirectory, false /* exploded */, monitor);
   }
 
-  private static void publish(IProject project, IPath destination, boolean exploded,
-      IProgressMonitor monitor) throws CoreException {
+  private static void publish(IProject project, IPath destination, IPath safeWorkDirectory,
+      boolean exploded, IProgressMonitor monitor) throws CoreException {
     if (monitor.isCanceled()) {
       throw new OperationCanceledException();
     }
     Preconditions.checkNotNull(project, "project is null"); //$NON-NLS-1$
     Preconditions.checkNotNull(destination, "destination is null"); //$NON-NLS-1$
     Preconditions.checkArgument(!destination.isEmpty(), "destination is empty path"); //$NON-NLS-1$
+    Preconditions.checkNotNull(safeWorkDirectory, "safeWorkDirectory is null"); //$NON-NLS-1$
 
     SubMonitor progress = SubMonitor.convert(monitor, 100);
     progress.setTaskName(Messages.getString("task.name.publish.war"));
 
-    PublishHelper publishHelper = new PublishHelper(null);
-    IModuleResource[] resources = flattenResources(project, progress);
+    IModuleResource[] resources = flattenResources(project, safeWorkDirectory, progress);
 
     if (exploded) {
```

Comment:
I think you can refactor to avoid the if and reduce the cyclomatic complexity. Return resources from this method and call publishFull or publishExploded from publishWar or publishExploded. Then the exploded argument and the if go away.

Reply:
> Return resources from this method and call publishFull or publishExploded from publishWar or publishExploded.

Will do so.

Reply:
Turns out this method does so little without the `if` statement and it just became `flatternResources` itself. Done.